### ISSUE TITLE
[TypeScript] Configurable Date type mapping

### ISSUE THREAD

**Issue by kernwig** [swagger-api/swagger-codegen#5141]
##### Description

Feature request to be able to configure the Date type mapping to use string instead of Javascript Date.

As of v2.3.0, codegen is doing the correct default behavior of translating a swagger date-time to a Javascript Date object. However, we all know that JS Date is deeply flawed, only really working for UTC. To do real date and time work, a 3rd party library such as js-joda must be used. If the APIs take the ISO 8601 format on the wire and put them into a Date object though, information is already lost.

##### Swagger-codegen version

2.3.0

##### Related issues

#4812 

##### Suggest a Fix

I'm not sure how to do configuration options. Currently, I'm forcing this with a modified source code:

```
@@ -42,7 +42,7 @@ public class TypeScriptAngular2ClientCodegen extends AbstractTypeScriptClientCod
         modelTemplateFiles.put("model.mustache", ".ts");
         apiTemplateFiles.put("api.service.mustache", ".ts");
         languageSpecificPrimitives.add("Blob");
-        typeMapping.put("Date","Date");
+        typeMapping.put("Date","string");
         typeMapping.put("file","Blob");
         apiPackage = "api";
         modelPackage = "model";
```

**Comment by kernwig** [comment]
Tried `--type-mappings Date=string`. That does what my code change above does... but the output of these parameters in angular2 is still `<any>endDate.toISOString()`.

Why the <any> cast when your clearly calling a Date function, on a string in this case?

**Comment by wing328** [comment]
@kernwig one way is to customize the templates and pass via the `-t` option.

Would using [Moment.js](https://momentjs.com/) instead of JS Date solve the problem for you?

**Comment by kernwig** [comment]
As I understand it, customizing templates means that I have to maintain my own templates forever... probably keeping up with any changes made to the project. That's not a good solution.

I don't see Moment.js being of any use. I believe that libraries works with strings and JS Date. The dates I/O as strings. We're good there. As soon as they become JS Date, they loose time zone information; there's no way to recover from there. Simply leave them as strings and let the user of the library do whatever they want - Date, moment, JodaTime... (I use JodaTime.)

I'm worked around this problem by writing a bash script that pre-processes the Swagger JSON spec and removes all the "format: date-time" properties. That leaves everything as strings and I'm happy. Everyone shouldn't have to do that though.

**Comment by wing328** [comment]
> As soon as they become JS Date, they loose time zone information

Yes, I'm aware of the limitations of JS Date. That's why I propose Moment.js, which seems to be able to handle datetime better across different browsers.

I agree with you that post-processing the auto-generated code is not ideal and we want to provide better datetime support for JS, TS generators out of the box.

**Comment by kernwig** [comment]
I've only used Moment.js as a conversion utility, not as actual types to carry data around. It doesn't seem to work for that, but maybe I'm just missing it.

js-joda is a library with date, time, duration, period... full on ISO-8601. However, it currently has a huge gaping hole when it comes to localization. (https://github.com/js-joda/js-joda/issues/34)

**Comment by wing328** [comment]
> As I understand it, customizing templates means that I have to maintain my own templates forever... probably keeping up with any changes made to the project. That's not a good solution.

Using `-t` can confirm if the solution works for you. Then feel free to submit a PR to add back the enhancement/fix to the project.

**Comment by kernwig** [comment]
I don't know off hand how to implement this request, it would take research. I was very easily able to write a pre-processor to modify the swagger spec file to my specific needs before it is processed. Chaining processing layers is a tried and true method. (i.e. The entire javascript ecosystem.) I'm fine with that. I haven't closed this request because I do think it is a nice-to-have for the project, but I don't have any pressing need for it myself anymore.

```
'use strict';

/*
 * Reads swagger definition file from backend and makes some modifications to it before generating the
 * client-side API with swagger-codegen.
 *
 * 1. Remove format:date-time from GET request parameters. We want these to be simple strings.
 *
 */
var https = require('https');

let url = process.argv[2];

https.get(url, (response) => {
    if (response.statusCode !== 200) {
        console.error(response.statusMessage);
        return;
    }
    else if (response.headers['content-type'].indexOf('json') < 0) {
        console.error("Bad content-type:" + response.headers['content-type']);
        return;
    }

    response.setEncoding('utf8');
    let rawData = '';
    response.on('data', (chunk) => rawData += chunk);
    response.on('end', () => {
        try {
            let parsedData = JSON.parse(rawData);
            let output = transform(parsedData);
            console.log(JSON.stringify(output));
        } catch (e) {
            console.error(e.message);
        }
    });
}).on('error', (e) => {
    console.error(`Got error: ${e.message}`);
});

function transform(def) {
    if (def && def.swagger) {
        for (let pathKey in def.paths) {
            if (!def.paths.hasOwnProperty(pathKey))
                continue;

            let fn = def.paths[pathKey].get;
            if (fn && fn.parameters) {
                for (let parameter of fn.parameters) {
                    if (parameter.type === 'string' && parameter.format === 'date-time') {
                        delete parameter.format;
                    }
                }
            }
        }
    }

    return def;
}
```

**Comment by macjohnny** [comment]
the type mapping of model properties can be configured as suggested in https://github.com/swagger-api/swagger-codegen/issues/5587#issuecomment-368805748
